coreneuron/network/partrans.cpp

```diff
@@ -41,40 +41,39 @@ void nrnmpi_v_transfer() {
     // gather the source values. can be done in parallel
     for (int tid = 0; tid < nrn_nthread; ++tid) {
         auto& ttd = transfer_thread_data_[tid];
-        auto& nt = nrn_threads[tid];
+        auto* nt = &nrn_threads[tid];
         int n = int(ttd.outsrc_indices.size());
         if (n == 0) {
             continue;
         }
-        double* src_data = nt._data;
+        double* src_data = nt->_data;
         int* src_indices = ttd.src_indices.data();
 
         // gather sources on gpu and copy to cpu, cpu scatters to outsrc_buf
         double* src_gather = ttd.src_gather.data();
         size_t n_src_gather = ttd.src_gather.size();
-        // clang-format off
 
-        #pragma acc parallel loop present(                                          \
-            src_indices[0:n_src_gather], src_data[0:nt._ndata],                     \
-            src_gather[0 : n_src_gather]) /*copyout(src_gather[0:n_src_gather])*/   \
-            if (nt.compute_gpu) async(nt.stream_id)
+        nrn_pragma_acc(parallel loop present(src_indices [0:n_src_gather],
+                                             src_data [0:nt->_ndata],
+                                             src_gather [0:n_src_gather]) if (nt->compute_gpu)
+                           async(nt->stream_id))
+        nrn_pragma_omp(target teams distribute parallel for simd if(nt->compute_gpu))
         for (int i = 0; i < n_src_gather; ++i) {
             src_gather[i] = src_data[src_indices[i]];
         }
-        // do not know why the copyout above did not work
-        // and the following update is needed
-        #pragma acc update host(src_gather[0 : n_src_gather])   \
-            if (nrn_threads[0].compute_gpu)                     \
-            async(nt.stream_id)
-        // clang-format on
+        nrn_pragma_acc(update host(src_gather [0:n_src_gather]) if (nt->compute_gpu)
+                           async(nt->stream_id))
+        nrn_pragma_omp(target update from(src_gather [0:n_src_gather]) if (nt->compute_gpu))
     }
 
     // copy gathered source values to outsrc_buf_
+    bool compute_gpu = false;
     for (int tid = 0; tid < nrn_nthread; ++tid) {
-        // clang-format off
-
-        #pragma acc wait(nrn_threads[tid].stream_id)
-        // clang-format on
+        if (nrn_threads[tid].compute_gpu) {
+            compute_gpu = true;
+            nrn_pragma_acc(wait(nrn_threads[tid].stream_id))
+            nrn_pragma_omp(taskwait)
```

Comment:
Olli, could you remind about this omp task wait requirement? I assume this is not absolute requirement today in OpenMP implementation. Is that true? (to be sure that I haven't missed this aspect at other places)

Reply:
I meant to drop all the `taskwait` directives when I dropped the first draft of the async support. I think I just missed this one. I'll delete it when I update the NMODL submodule.